### PR TITLE
added the tagging module

### PR DESCRIPTION
**Pull request by paznewman**

### CODE REVIEW

File: tagging.py [hunk 1]
```diff
@@ -0,0 +1,185 @@
+from mutagen import File
+from mutagen.easymp4 import EasyMP4 # used for .m4a and .mp4 files
+from mutagen.flac import FLAC
+from mutagen.easyid3 import EasyID3 # used for .mp3 files
+
```
Comment (Omerbu) [inline tagging.py:5]: module docs come before imports
PEP 257

File: tagging.py [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+# the AUDIO_TYPES constant dict is used to refer to the mutagen subclasses according to the file format
+#
+# the TAGS_MAPS constant dict is used to refer to the correct tag names for each file format.
+# currently all tags appear the same. that is because the use of EasyID3 and EasyMP4, which mimic the APEv2 format. 
+# regardless, this structure is kept for probable future use.
+AUDIO_TYPES = {"mp3":EasyID3, "mp4":EasyMP4, "flac":FLAC}
+TAGS_MAPS = {'mp3':{'artist':'artist','album':'album','title':'title','track':'tracknumber','year':'date','genre':'genre','albumartist':'albumartist'}, # mp3 tags are EasyID3
```
Comment (Omerbu) [inline tagging.py:12]: PEP8
Line too long

File: tagging.py [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+             'mp4':{'artist':'artist','album':'album','title':'title','track':'tracknumber','year':'date','genre':'genre','albumartist':'albumartist'}, # mp4 tags are EasyMP4
+             'flac':{'artist':'artist','album':'album','title':'title','track':'tracknumber','year':'date','genre':'genre','albumartist':'albumartist'}}
+
+
+class AudioFile:
```
Comment (Omerbu) [inline tagging.py:17]: every Class should inherit from "object" class.
And "AudioFile" is too generic and might cause a namespace mixing.

File: tagging.py [hunk 4]
```diff
@@ -0,0 +1,185 @@
+from mutagen import File
+from mutagen.easymp4 import EasyMP4 # used for .m4a and .mp4 files
+from mutagen.flac import FLAC
+from mutagen.easyid3 import EasyID3 # used for .mp3 files
+
+# the AUDIO_TYPES constant dict is used to refer to the mutagen subclasses according to the file format
+#
+# the TAGS_MAPS constant dict is used to refer to the correct tag names for each file format.
```
Comment (Omerbu) [inline tagging.py:8]: USE PEP257 docstring conventions

File: tagging.py [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+AUDIO_TYPES = {"mp3":EasyID3, "mp4":EasyMP4, "flac":FLAC}
+TAGS_MAPS = {'mp3':{'artist':'artist','album':'album','title':'title','track':'tracknumber','year':'date','genre':'genre','albumartist':'albumartist'}, # mp3 tags are EasyID3
+             'mp4':{'artist':'artist','album':'album','title':'title','track':'tracknumber','year':'date','genre':'genre','albumartist':'albumartist'}, # mp4 tags are EasyMP4
+             'flac':{'artist':'artist','album':'album','title':'title','track':'tracknumber','year':'date','genre':'genre','albumartist':'albumartist'}}
+
+
+class AudioFile:
+    def __init__(self,filename):
+        audiotype = File(filename).mime[0].split("/")[1]    # determine the file format
+        self.meta = AUDIO_TYPES[audiotype](filename)        # create the tags instance of the appropriate mutagen subclass
```
Comment (Omerbu) [inline tagging.py:20]: uninformative name

File: tagging.py [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+    
+    def delete_tags(self):
+        '''
+            remove all the tags from the file.
+            this leaves the tags as an empty dict. 
+        '''
+        self.meta.clear()
+        self.meta.save()
+    
+    def artist(self,entry=None):
```
Comment (Omerbu) [inline tagging.py:42]: bad naming in all of setting methods
"set_artist/name/album_tag" would be a lot better

File: tagging.py [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+        print 'Title:       ' + self.meta[self.tag_map['title']][0]
+        print 'Track:       ' + self.meta[self.tag_map['track']][0]
+        print 'Year:        ' + self.meta[self.tag_map['year']][0]
+        print 'Genre:       ' + self.meta[self.tag_map['genre']][0]
+        print 'AlbumArtist: ' + self.meta[self.tag_map['albumartist']][0]
+
+
+def edit_tags(filename,tags_dict):
+    '''
+       receive a file name to edit and the newly fetched tags.
```
Comment (Omerbu) [inline tagging.py:128]: bad doc

File: tagging.py (hunk 6 above)
Comment (paznewman) [inline tagging.py:42]: these functions are used both for setting and getting

File: tagging.py (hunk 5 above)
Comment (paznewman) [inline tagging.py:20]: meta for metadata. 
self.tags seems confusing

File: tagging.py [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+            this leaves the tags as an empty dict. 
+        '''
+        self.meta.clear()
+        self.meta.save()
+    
+    def artist(self,entry=None):
+        '''
+            return or set the file's artist
+        '''
+        if entry:
```
Comment (Omerbu) [inline tagging.py:46]: massive code duplication. (bad!)

File: tagging.py [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+        if audio.meta[audio.tag_map[tag]] and tags2compare[tag]:
+            result += _compare_words(tags2compare[tag],audio.meta[audio.tag_map[tag]][0])  
+        else:
+            result += 0
+    return result / len(TAGS_MAPS['mp3'])
+    
+    
+
+def check_tags(filename):
+    '''
```
Comment (Omerbu) [inline tagging.py]: Should be sepreated to a unittest

File: tagging.py [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+        returns a float between 0 to 1 as a result, where 1 is a perfect match.
+    '''
+    def _compare_words(word1, word2):
+        c = 0.0     # initiated as float for the last division
+        str1 = word1.replace(" ","").lower()    # ignore all spaces and case
+        str2 = word2.replace(" ","").lower()    # ignore all spaces and case
+        if len(str1) != len(str2): 
+            # assign the longer string to str1
+            str1, str2 = max([str1,str2],key=len), min([str1,str2], key=len)
+        for i in range(len(str1)):
```
Comment (Omerbu) [inline tagging.py:152]: use xrange

File: tagging.py [hunk 11]
```diff
@@ -0,0 +1,185 @@
...
+
+def compare_tags(filename, tags2compare):
+    '''
+        should be used as a last resort for checking best match of new tags.
+        
+        get a dict of new tags and compare each tag to the corresponding old tag
+        returns a float between 0 to 1 as a result, where 1 is a perfect match.
+    '''
+    def _compare_words(word1, word2):
+        c = 0.0     # initiated as float for the last division
```
Comment (Omerbu) [inline tagging.py:146]: Uninformative variable name
use variable = float(), not = 0.0

File: tagging.py [hunk 12]
```diff
@@ -0,0 +1,185 @@
...
+            str1, str2 = max([str1,str2],key=len), min([str1,str2], key=len)
+        for i in range(len(str1)):
+            if i < len(str2) and str1[i] == str2[i]:
+                c += 1
+            else:
+                c += 0
+        return c / len(str1)
+    
+    audio = AudioFile(filename)
+    result = 0.0    # initiated as float for the last division
```
Comment (Omerbu) [inline tagging.py:160]: see above about 0.0

File: tagging.py [hunk 13]
```diff
@@ -0,0 +1,185 @@
...
+        if entry:
+            entry = str(entry)
+            self.meta[self.tag_map['albumartist']] = [entry.decode("UTF-8")]
+            self.meta.save()
+        return self.meta[self.tag_map['albumartist']][0]
+    
+    def tag_print(self):
+        """
+            nicely prints the tags. 
+            should probably be removed when the project is finished
```
Comment (Omerbu) [inline tagging.py:115]: part of unittest or debug module.

File: tagging.py [hunk 14]
```diff
@@ -0,0 +1,185 @@
...
+        self.meta.save()
+    
+    def artist(self,entry=None):
+        '''
+            return or set the file's artist
+        '''
+        if entry:
+            entry = str(entry)
+            self.meta[self.tag_map['artist']] = [entry.decode("UTF-8")]
+            self.meta.save()
```
Comment (Omerbu) [inline tagging.py:49]: Why not:
``def set_tag(self, tag_type. tag_data):
     if tag_data:
        tag_data = str(tag_data)
        self.meta[self.tag_map[tag_type]] = tag_data .......

File: tagging.py [hunk 15]
```diff
@@ -0,0 +1,185 @@
...
+    '''
+    audio = AudioFile(filename)
+    audio.tag_print()
+    
+    test  ={'artist':'Cat Stevens','album':'Tea For The Tillerman','title':'Sad Lisa','track':'4','year':'1970','genre':'Folk Rock','albumartist':'Cat Stevens'}
+    print compare_tags(filename, test)
+    
+    return audio.meta
+
+if __name__ == '__main__':
```
Comment (Omerbu) [inline tagging.py:183]: A class module doesn't need a "if dunder name == dunder main"

File: tagging.py [hunk 16]
```diff
@@ -0,0 +1,185 @@
...
+# regardless, this structure is kept for probable future use.
+AUDIO_TYPES = {"mp3":EasyID3, "mp4":EasyMP4, "flac":FLAC}
+TAGS_MAPS = {'mp3':{'artist':'artist','album':'album','title':'title','track':'tracknumber','year':'date','genre':'genre','albumartist':'albumartist'}, # mp3 tags are EasyID3
+             'mp4':{'artist':'artist','album':'album','title':'title','track':'tracknumber','year':'date','genre':'genre','albumartist':'albumartist'}, # mp4 tags are EasyMP4
+             'flac':{'artist':'artist','album':'album','title':'title','track':'tracknumber','year':'date','genre':'genre','albumartist':'albumartist'}}
+
+
+class AudioFile:
+    def __init__(self,filename):
+        audiotype = File(filename).mime[0].split("/")[1]    # determine the file format
```
Comment (Omerbu) [inline tagging.py:19]: Should be instance (file) specific, and therfore "self.audiotype".

**Pull request closed without merging**